blogposts/2020/performance-improvements-in-precise-code-intel.md

```diff
@@ -0,0 +1,256 @@
+
+The correlation process does little to no I/O: it just inserts identifiers into the proper map, slice, and set structures for the next step of processing. Profiling showed that the most CPU time at this phase of processing was spent in the `"encoding/json"` package (due to its heavy use of reflection). Similarly, the highest number of allocations were also from decoding values into structs (followed closely by serialization of different structs on write).
+
+The input we accept is not easily changeable: the LSIF protocol defines the output format, and we'd like to be able to accept any protocol-confirming indexer output. It's a non-starter to require a Sourcegraph-specific input format, which would limit the available indexers to those that are written specifically for Sourcegraph and break backwards compatibility with existing users.
+
+Instead, we looked into alternative libraries to parse JSON input. We chose [json-iterator/go](https://github.com/json-iterator/go), which is a high-performance drop-in replacement. Switching to this library was a very minimal change and gave a very nice speed boost. This library is particular faster at decoding small structures, and most LSIF vertex and edge definitions are very small (with the exception of some high-degree _contains_ edges, as is the case with very long files).
+
+Alternative libraries we considered include [easyjson](https://github.com/mailru/easyjson), [fastjson](https://github.com/valyala/fastjson), and [ffjson](https://github.com/pquerna/ffjson) among a few others. Some of these libraries required additional engineering effort as the API was not equivalent to the API defined by the `"encoding/json"` package.
+
+### Reduce disk writes
+
+**PR**: [codeintel: GetUpload should return reader, not file (#11042)](https://github.com/sourcegraph/sourcegraph/pull/11042)
+
+The bundle manager is the keeper of all things persistent in the precise code intel world. When a worker needs to process raw LSIF data, it needs to request the data from the bundle manager. Symmetrically, once a worker serializes a bundle, it is passed back to the bundle manager for permanent storage.
+
+Prior to this change, the bundle manager client would request the data via an HTTP request, write it to disk, then pass the filename to the worker process. The worker would open and read the file for processing.
+
+We can reduce I/O by simply passing the HTTP response reader back to the worker instead of requiring that it hit disk. This also saves us the time required to wait for the entire transfer to complete and flush to disk before beginning to process the data. As raw LSIF indexes can be quite large (multi-gigabyte), this provides a non-negligible boost in many cases.
```

Comment:
Same here — this ending feels a bit abrupt!

Reply:
Agreed. Same as above - any suggestions here? (This is an open question to anyone on the team 😄 ).

Reply:
The charts at the top show the overall improvement, right? Not sure anything specific is needed here — maybe just a pitch for "Interested in what we're doing here? We're hiring!"

Reply:
Updated.